src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Finalization.cs

```diff
@@ -57,5 +57,54 @@ public async Task IncompleteResponseStream_ResponseDropped_CancelsRequestToServe
                 }));
             }
         }
+
+        [Fact]
+        public async Task ClientNotDisposed_ServerReadsFromStream_ServerHangs()
+        {
+            Diagnostics.Debugger.Launch();
+            using (HttpClient client = new HttpClient())
+            {
+                await LoopbackServerFactory.CreateClientAndServerAsync(async url =>
+                {
+                    var stream = await client.GetStreamAsync(url);
+                    stream.Close();
+                    //client.Dispose(); Uncomment this to close the request stream and make the test succeed
```

Comment:
Closing the stream returned by `GetStreamAsync` does not mean it closes the socket -- it only closes that single request. HTTP/1.1 with keepalive, or HTTP/2 in this case, re-uses a connection for multiple requests, so `HttpClient` will keep it alive.

You're reading directly from the `Http2Connection`'s stream, which will give you raw HTTP2 protocol rather than request content. So you're breaking the protocol here (use `ReadBodyAsync` instead) and because `HttpClient` doesn't close the socket, `Stream.Read` will just block until it sees another request.

Reply:
> You're reading directly from the Http2Connection's stream, which will give you raw HTTP2 protocol rather than request content.

That's interesting. I would have expected `GetStreamAsync` to throw a NotSupportedException for HTTP2.

Reply:
> I would have expected GetStreamAsync to throw a NotSupportedException for HTTP2.

Why?  GetStreamAsync gives you back a Stream to let you read the request's response.  GetByteArrayAsync, for example, effectively just does GetStreamAsync, reads to EOF, and returns all the bytes it read.

Reply:
Nevermind, I misunderstood the original statement.

Reply:
Also note that the client and server streams are very different. One represents response context. On HTTP2m, you can have multiples at the same time as we can multiplex.
The server stream from LoopbackServer is truly  IO stream without any HTTP knowledge. So the read will hung until connection is closed by client.  

You can demonstrate this and force this by modifying code like:
```c#
            using (var handler = new SocketsHttpHandler())
            using (HttpClient client = new HttpClient(handler))
            {
                handler.PooledConnectionLifetime = TimeSpan.FromMilliseconds(100);
                handler.PooledConnectionIdleTimeout = TimeSpan.FromMilliseconds(100);
```

With this, connection will be retired without disposing client and that will wake up server stream.

Reply:
Sorry, I forgot to mention that all testing for was done for **HTTP 1.1 only** by running SocketsHttpHandler_HttpClientHandler_Finalization_Http11_Test.ClientNotDisposed_ServerReadsFromStream_ServerHangs.
I understand the point that in the given case the server will wait for the next byte to arrive on the server stream until it's closed. This is expected. The surprising point here is the reason why the server stream gets closed. It's expected that it gets closed when client connection times out. However, it's unexpected that the server stream gets closed when HttpClient gets disposed because I believe HttpClient returns the connection back to the pool after the current request completes and it's not supposed to hold any references to connection-related resources.

Reply:
> However, it's unexpected that the server stream gets closed when HttpClient gets disposed because I believe HttpClient returns the connection back to the pool after the current request completes and it's not supposed to hold any references to connection-related resources.

This is not correct. HttpClient (its underlying handler) owns the connection pool. It will close of all connections in the pool when it's disposed.

Reply:
I see, then it's by design and there is no issue. Thank you for clarification.